aten/src/ATen/native/cpu/UpSampleMoreKernel.cpp

```diff
@@ -0,0 +1,575 @@
+#include <ATen/ATen.h>
+
+#include <ATen/Dispatch.h>
+#include <ATen/native/UpSample.h>
+#include <ATen/Parallel.h>
+#include <ATen/cpu/vec256/vec256.h>
+
+namespace at {
+namespace native {
+namespace {
+
+template<typename scalar_t>
+static inline void compute_source_index_and_lambda(
+    int64_t& input_index0,
+    int64_t& input_index1,
+    scalar_t& lambda0,
+    scalar_t& lambda1,
+    scalar_t ratio,
+    int64_t output_index,
+    int64_t input_size,
+    int64_t output_size,
+    bool align_corners) {
+  if (output_size == input_size) {
+    // scale_factor = 1, simply copy
+    input_index0 = output_index;
+    input_index1 = output_index;
+    lambda0 = static_cast<scalar_t>(1);
+    lambda1 = static_cast<scalar_t>(0);
+  } else {
+    const scalar_t real_input_index = area_pixel_compute_source_index<scalar_t>(
+        ratio, output_index, align_corners, /*cubic=*/false);
+    input_index0 = static_cast<int64_t>(real_input_index);
+    int64_t offset = (input_index0 < input_size - 1) ? 1 : 0;
+    input_index1 = input_index0 + offset;
+    lambda1 = real_input_index - input_index0;
+    lambda0 = static_cast<scalar_t>(1.) - lambda1;
+  }
+}
+
+template <typename scalar_t, typename scale_type>
+void cpu_upsample_linear(
+    Tensor& output_,
+    const Tensor& input_,
+    bool align_corners,
+    const scale_type& scales) {
+  TORCH_CHECK(input_.dtype() == output_.dtype(), "expected dtype ", input_.dtype(),
+              " for `output` but got dtype ", output_.dtype());
+  auto input = input_.contiguous();
+  auto output = output_.contiguous();
+
+  auto input_data = input.data_ptr<scalar_t>();
+  auto output_data = output.data_ptr<scalar_t>();
+  auto input_sizes = input.sizes().vec();
+  auto output_sizes = output.sizes().vec();
+  auto ndim = input_sizes.size();
+  auto numel = output.numel();
+
+  // treat nbatch and channels as one dimension
+  int64_t channels = input_sizes[0] * input_sizes[1];
+  int64_t input_depth = (ndim == 5) ? input_sizes[2] : 1;
+  int64_t output_depth = (ndim == 5) ? output_sizes[2] : 1;
+  int64_t input_height = (ndim >= 4) ? input_sizes[ndim - 2] : 1;
+  int64_t output_height = (ndim >= 4) ? output_sizes[ndim - 2] : 1;
+  int64_t input_width = input_sizes[ndim - 1];
+  int64_t output_width = output_sizes[ndim - 1];
+
+  int64_t output_slice_size = output_depth * output_height * output_width;
+
+  auto loop1d = [&](int64_t begin, int64_t end) {
+    const scalar_t width_scale = area_pixel_compute_scale<scalar_t>(
+        input_width, output_width, align_corners, scales[0]);
+
+    auto input_indexr = [=](int64_t c, int64_t w) {
+      return input_data[c * input_width + w];
+    };
+
+    int64_t iw0, iw1;
+    scalar_t w0lambda, w1lambda;
+    for (int64_t c = begin; c < end; c++) {
+      for (int64_t ow = 0; ow < output_width; ow++) {
+        compute_source_index_and_lambda(
+            iw0, iw1, w0lambda, w1lambda, width_scale, ow, input_width, output_width, align_corners);
```

Comment:
Do you guys care about bfloat16 here? Computing w?lambda coefficients in low precision may not be enough, but it can be left as a followup.

Reply:
Yes, we do care about bf16. Let us leave bf16 as a follow up to align with bigger scope.